Replace a custom logger with off the shelf implementation (#7559)
enso-org/enso: docs/infrastructure/logging.md

The Enso project features a centralised logging service to allow for the
aggregation of logs from multiple components. This service can be started with
one of the main components, allowing other components to connect to it. The
service aggregates all logs in one place for easier analysis of the interaction

> between components. Components can also write to console, files directly without

Reviewer: ```suggestion
between components. Components can also log to console or files directly without
```

Reviewer: Programs may of course print to console or write to files, but what does that have to do with logging? That was my first question after reading this sentence.

> involving an additional logging server.

Reviewer: ```suggestion
involving the centralized logging service.
```

Reviewer: Refers to line 11 which mentions the "centrailized logging service".

- [Configuration](#configuration)
  - [Custom Log Levels](#custom-log-levels)
  - [Appenders](#appenders)
    - [Format](#format)
    - [File](#file-appender)
    - [Network](#socket-appender)
    - [Sentry.io](#sentry-appender)
- [JVM Architecture](#jvm-architecture)
  - [SLF4J Interface](#slf4j-interface)
  - [Setting Up Logging](#setting-up-logging)
  - [Log Masking](#log-masking)
  - [Logging in Tests](#logging-in-tests)

<!-- /MarkdownTOC -->

## Configuration

The logging settings should be placed under the `logging-service` key of the
`application.conf` config. Each of the main components can customize format and
output target via section in `application.conf` configuration file. The
configuration is using HOCON-style, as defined by
[lightbend/config](https://github.com/lightbend/config). Individual values
accepted in the config are inspired by SLF4J's properties, formatting and
implementations.

The configuration has two main sections:

- [custom log levels](#custom-log-levels)
- [applications' appenders](#appenders) (also known as configuration of log
  events output target)

During component's setup, its `application.conf` file is read, parsed and

> verified and available as part of the class hierarchy defined in

Reviewer: > verified and available

I don't understand.

`org.enso.logger.config` package. Class
`org.enso.logger.config.LoggingServiceConfig` encapsulates the complete
information represented by the `logging-service` key of the config file and is
used to programmatically initialize loggers.

As per [configuration schema](https://github.com/lightbend/config) any key can

> have a default value that can be overridden by an environmental variable. For

Reviewer 2: ```suggestion
have a default value that can be overridden by an environment variable. For
```

nitpick, but I have never heard `environmental variable` before; sounds like a term related to ecology rather than programming 😅

example

```
  {
    host = localhost
    host = $ENSO_HOST
  }
```

defines a `host` key once, except that `ENSO_HOST` values takes a precedence if
it is defined during loading of the config file.

### Custom Log Levels

The `logging-service.logger` configuration provides an ability to override the
default application log level for particular loggers. In the `logger` subconfig
the key specifies the logger name (or it's prefix) and the value specifies the
log level for that logger.

```
logging-service.logger {
  akka.actor = info
  akka.event = error
  akka.io = error
  slick {
    jdbc.JdbcBackend.statement = debug
    "*" = error
  }
}
```

For example, the config above limits all `akka.actor.*` loggers to the info
level logging, and `akka.event.*` loggers can emit only the error level
messages.

Config supports globs (`*`). For example, the config above sets
`jdbc.JdbcBackend.statement` SQL statements logging to debug level, and the rest

> of the slick loggers to error level.

Reviewer: What is the FQN of the slick loggers? Is it `slick.jdbc.JdbcBackend.statement`? What is the difference between setting:
```
slick = error
slick.jdbc.JdbcBackend.statement = debug
```
and using the above `"*"` form?

Author: Order matters. 
```
slick = error
slick.jdbc.JdbcBackend.statement = debug
```
will always  match `slick` classes.

I guess your question was more
```
  slick.jdbc.JdbcBackend.statement = debug
  slick = error
```
vs
```
  slick.jdbc.JdbcBackend.statement = debug
  slick.* = error
```

Well, no diff.

Reviewer: Then I don't understand why (it is important to stress that) _"Config supports globs (`*`)"_. If simple `slick = error` does the trick, then why advertise the complex syntax? Probably more important to stress that order matters.

Additionally, custom log events can be provided during runtime via system
properties, without re-packaging the updated config file. For example

```typescript
akka.actor = info;
```

is equivalent to

```typescript
  -Dakka.actor.Logger.level=info
```

> Any custom log level is therefore defined with `-Dx.y.Z.Logger.level` where `x`,

Reviewer: +1

`y` and `Z` refer to the package elements and class name, respectively. System
properties always have a higher priority over those defined in the
`application.conf` file.

### Appenders

Log output target is also configured in the `application.conf` files in the
"appenders" section ("appender" is equivalent to `java.util.logging.Handler`
semantics). Each appender section can provide further required and optional

> key/value pairs, to better customize the log target output.

Reviewer: At this point of document, I'd like to know how one chooses the each appender?

Currently supported are

- console appender - the most basic appender that prints log events to stdout
- [file appender](#file-appender) - appender that writes log events to a file,
  with optional rolling file policy
- [socket appender](#socket-appender) - appender that forwards log events to
  some logging server

> - [sentry.io appender](#sentry-appender) - appender that forwards log events to

Reviewer: Good, support for `sentry.io` should be interesting for @PabloBuchu.

  a sentry.io service

The appenders are defined by the `logging-service.appenders`. Currently only a
single appender can be selected at a time. The selection may also be done via an

> environmental variable `$ENSO_DEFAULT_APPENDER`.

Reviewer 2: Would it be possible to still have an stdout appender when selecting the file appender? While file appender is very useful, I think it's good to be able to still see the logs in the terminal.

We can do `tail -f logfile` but allowing the app to do this automatically would be ideal.

Author: I was going to add support for multiple appenders in the follow up PR. Also, it would probably make sense to have different levels for different loggers.

#### Format

The pattern follows the classic's
[PatternLayout](https://logback.qos.ch/manual/layouts.html#ClassicPatternLayout)
format.

Appenders that store/display log events can specify the format of the log
message via `pattern` field e.g.

```typescript

  appenders = [
    {
      name = "console"
      pattern = "[%level{lowercase=true}] [%d{yyyy-MM-dd'T'HH:mm:ssXXX}] [%logger] %msg%n%nopex"
    }
    ...
  ]
```

#### File Appender

File appender directs all log events to a log file:

```
  {
    name = "file"
    append = <boolean, optional>
    immediate-flush = <boolean, optional>
    pattern = <string, optional>
    rolling-policy {
      max-file-size = <string, optional>
      max-history = <int, optional>
      max-total-size = <string, optional>
    }
  }
```

Rolling policy is a fully optional property of File Appender that would trigger
automatic log rotation. All properties are optional with some reasonable

> defaults if not specified.

Reviewer 2: Can we describe what are these 'some reasonable defaults'? Otherwise, where can I find these defaults?

#### Socket Appender

Configuration

```
  {
    name = "socket"
    hostname = <string, required>
    port = <string, required>
  }
```

The two fields can be overridden via environment variables:

> - `hostanme` has an equivalent `$ENSO_LOGSERVER_HOSTNAME` variable

Reviewer: ```suggestion
- `hostname` has an equivalent `$ENSO_LOGSERVER_HOSTNAME` variable
```

- `port` has an equivalent `$ENSO_LOGSERVER_PORT` variable

#### Sentry Appender

```
  {
    name = "sentry"

>     dsn = <string, required>

Reviewer: Shouldn't there be an environment variable to provide `dsn`? I assume @PabloBuchu would like to configure the value without changing engine/LS source code.

  }
```

Sentry's Appender has a single required field, `dsn`.

## JVM Architecture

Previously, Enso came with its own implementation of Logger. That has been
replaced with a mostly off-the-shelf logging implementation, avoiding

> maintenance costs for features that were already present in them.

Reviewer: There three sentences belong to a PR comment or description, but not to _"up-to-date"_ documentation. Who cares about _"previously"_?

### SLF4J Interface

The user code must not be calling any of the underlying implementations, such as
Log4J or Logback, and should only request loggers via factory methods.

One can use the `org.slf4j.LoggerFactory` directly to retrieve class-specific
logger. For Scala code, it is recommended to use the
`com.typesafe.scalalogging.Logger` instead which wraps the SLF4J logger with
macros that compute the log messages only if the given logging level is enabled,
and allows much prettier initialisation.

```java
package foo;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Foo {
    private Logger logger = LoggerFactory.getLogger(Foo.class);

    public void bar() {
        logger.info("Hello world!");
    }
}
```

### Setting Up Logging

The `org.slf4j.Logger` instances have to know where to send log events. This
setting is typically performed once, when the service starts, and applies
globally during its execution. Currently, it is not possible to dynamically
change where log events are being stored. The main class used for setting up
logging is `org.enso.logger.LoggerSetup` but it should not be instantiated

> directly by the users. Instead, it should be retrieved with the Thread-safe

Reviewer: `LoggerSetup` is abstract. It cannot be instantiated. Btw. in current form the `get` method isn't thread-safe - it just pretends to be...

`org.enso.logger.LoggerSetup.get` factory method. `org.enso.logger.LoggerSetup`
provides a number of `setupXYZAppender` methods that will direct loggers to send
log events to an `XYZ` appender. Setting a specific hard-coded appender
programmatically should however be avoided by the users. Instead, one should
invoke one of the overloaded `setup` variants that initialize loggers based on
the provided `logging-service` configuration.

```java
package foo;
import org.enso.logger.LoggerSetup;
import org.slf4j.event.Level;

public class MyService {

  private Logger logger = LoggerFactory.getLogger(Foo.class);
  ...
  public void start(Level logLevel) {
    LoggerSetup.get().setup(logLevel);
    logger.info("My service is starting...");
    ...
  }
  ...
}
```

`org.enso.logging.LoggingSetupHelper` class was introduced to help with the most
common use cases - establishing a file-based logging in the Enso's dedicated
directories or connecting to an existing logging server once it starts accepting
connections. That is why services don't call `LoggerSetup` directly but instead
provide a service-specific implementation of

> `org.enso.logging.LoggingSetupHelper`. `LoggingSetupHelper` and `LoggerSetup`

Reviewer: Isn't there an inconsistency in naming? `Logger`Setup, but `Logging`SetupHelper. I'd expect the same prefix.

because it uses default `toString` implementation and can leak critical
information even if the object implements custom interface for masked logging.

### Logging in Tests

The Logging Service provides a helper function `TestLogger.gatherLogs` that will
execute the closure and collect all logs reported in the specified class. That

> way it can verify that all logs are being reported within the provided code.

Reviewer 2: ❤️
